### PR TITLE
[WIP] fragments: init at 2.0-unstable

### LINKED ISSUE #132316: Packaging request: Fragments
**Project description**
Fragments is a modern, responsive and graphical Gnome bittorrent client

**Metadata**

* homepage URL: https://gitlab.gnome.org/World/Fragments
* source URL: https://gitlab.gnome.org/World/Fragments
* license: gpl3
* platforms: linux

The new version of Fragments depends on libutp which hopefully gets merged soon. I've opened a pull request for it [here](https://github.com/NixOS/nixpkgs/pull/132315).

[Earlier closed pull request](https://github.com/NixOS/nixpkgs/pull/56793) for Fragments.

### PR DESCRIPTION
**Pull request by onny**
###### Motivation for this change
Package for the Gnome torrent client [Fragments](https://gitlab.gnome.org/World/Fragments). It has a modern, simple and adaptive interface. Works well on desktop and mobile devices such as NixOS mobile smartphones.

Fixes https://github.com/NixOS/nixpkgs/issues/132316 but currently depends on libutp being merged https://github.com/NixOS/nixpkgs/pull/132315

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/p2p/fragments/default.nix [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+stdenv.mkDerivation rec {
+  pname = "fragments";
+  version = "1.5";
+
+  src = fetchgit {
+    url = "https://gitlab.gnome.org/World/Fragments";
+    rev = version;
+    fetchSubmodules = true;
+    sha256 = "0x1kafhlgyi65l4w67c24r8mpvasg3q3c4wlgnjc9sxvp6ki7xbn";
+  };
```
Comment (hrhino) [inline pkgs/applications/networking/p2p/fragments/default.nix:36]: can you perhaps use
```nix
fetchFromGitLab {
  domain = "gitlab.gnome.org";
  owner = "World";
  repo = "Fragments";
  rev = version;
  fetchSubmodules = true;
  sha256 = "0x1kafhlgyi65l4w67c24r8mpvasg3q3c4wlgnjc9sxvp6ki7xbn";
}
```

File: pkgs/applications/networking/p2p/fragments/find_library.patch [hunk 2]
```diff
@@ -0,0 +1,26 @@
+--- a/meson.build	2020-09-14 20:27:52.539520481 +0200
++++ b/meson.build	2020-09-14 20:27:59.023657234 +0200
+@@ -21,7 +21,7 @@
+ gio_dep = dependency('gio-2.0')
+ gio_unix_dep = dependency('gio-unix-2.0')
+ gtk_dep = dependency('gtk+-3.0', version : '>=3.20.10')
+-handy_dep = dependency('libhandy-1')
++#handy_dep = dependency('libglade-handy')
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/fragments/find_library.patch:8]:
```suggestion
```

File: pkgs/applications/networking/p2p/fragments/default.nix [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    chmod +x build-aux/*
+    patchShebangs build-aux
+  '';
+
+  dontUseCmakeConfigure = true;
+
+  meta = with lib; {
+    description = "An easy to use BitTorrent client which follows the GNOME HIG and includes well thought-out features";
+    homepage = https://gitlab.gnome.org/World/Fragments;
+    license = licenses.gpl3;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/fragments/default.nix:78]: Do you know if the license is gpl3Only or gpl3Plus?

File: pkgs/applications/networking/p2p/fragments/default.nix [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  nativeBuildInputs = [
+    autoconf
+    automake
+    cmake
+    desktop-file-utils
+    libtool
+    meson
+    ninja
+    pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/fragments/default.nix:46]:
```suggestion
    pkg-config
```

Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/fragments/default.nix:46]: You are sure we need all of these?

File: pkgs/applications/networking/p2p/fragments/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/applications/networking/p2p/fragments/default.nix:36]: fetchSubmodules only works with a very recent version of nixpkgs IIRC.

File: pkgs/applications/networking/p2p/fragments/default.nix [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+, curl
+, vala
+, glib
+, python3
+, gtk3
+, libhandy
+, hicolor-icon-theme
+, libtransmission
+, libb64
+, wrapGAppsHook }:
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/fragments/default.nix:25]:
```suggestion
, wrapGAppsHook
}:
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by onny** [comment]
I also added some small missing dependencies in this PR. Compilation seems to fail with linking:
```
/nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld: /nix/store/wriil72nvx3dyr136d02cgx4h2njgjr7-transmission-3.00/lib/libtransmission.a(crypto-utils-openssl.c.o): undefined reference to symbol 'SSL_CTX_get_cert_store@@OPENSSL_1_1_0'
/nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld: /nix/store/79lh7p2jjwqimzwksnpaclc3zj42zqda-openssl-1.1.1k/lib/libssl.so.1.1: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by onny** [comment]
Created an upstream issue https://gitlab.gnome.org/World/Fragments/-/issues/106

File: pkgs/development/libraries/dht/default.nix [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  #installPhase = ''
+  #  mkdir -p $out/lib $out/include/dht
+  #  cp dht.a dht.so $out/lib/
+  #  cp *.h $out/include/dht/
+  #'';
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/dht/default.nix:26]: What is with this?

File: pkgs/applications/networking/p2p/fragments/default.nix (hunk 4 above)
Comment (onny) [reply pkgs/applications/networking/p2p/fragments/default.nix]: Good point, I'll check this after I get this to compile properly :)

**Comment by onny** [comment]
2021-08-09-145452_grim

Got it to compile and run successfully :) Now I'll have to figure out which dependencies are still required ...

File: pkgs/applications/networking/p2p/fragments/default.nix [hunk 7]
```diff
@@ -89,7 +94,12 @@ rustPlatform.buildRustPackage rec {
     cargo
   ];
 
-  #patches = [ ./find_library.patch ];
+  # Don't use buildRustPackage phases, only use it for rust deps setup
+  configurePhase = null;
+  buildPhase = null;
+  doCheck = true;
+  checkPhase = null;
+  installPhase = null;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/fragments/default.nix:95]: Please use the don't variants instead.

**Comment by angustrau** [comment]
Sorry I didn't see this PR before starting on your packaging request! I'm happy to close mine (#134007) if this fits your use case better :)

**Comment by onny** [comment]
If this stable version is working, we should merge this one since my PR is for unstable version 2.0 and I'm not sure when it gets released :)

**Comment by BlueManCZ** [comment]
Hello @onny, how did you manage to fix the [`libssl symbol error`](https://gitlab.gnome.org/World/Fragments/-/issues/106)? I'm trying to build Fragments on Gentoo and got into the exact same issue.

**Comment by onny** [comment]
@BlueManCZ Have you tried using my package? Last time I build it, it was working fine (with 21.05 channel)

**Comment by BlueManCZ** [comment]
No, I haven't. I don't have NixOS. I only found your upstream issue which brought me here. But I think adding `ssl` dependency resolved that. I will ask Gentoo devs for help because this another problem seems to be related to the `libtransmission.a` static library and its compilation.

**Pull request closed without merging**